PortalSwift/Sources/PortalSwift/Storage/ICloudStorage.swift

```diff
@@ -0,0 +1,116 @@
+//
+//  ICloudStorage.swift
+//
+//  Created by Portal Labs, Inc.
+//  Copyright © 2022 Portal Labs, Inc. All rights reserved.
+//
+
+import CommonCrypto
+import Foundation
+
+/// A storage class that uses iCloud's key-value store to store/retrieve private keys.
+public class ICloudStorage: Storage, PortalStorage {
+  public enum ICloudStorageError: Error {
+    case noAPIKeyProvided(String)
+    case noAccessToICloud(String)
+    case notSignedIntoICloud(String)
+    case unableToDeriveKey(String)
+    case unableToRetrieveClient(String)
+    case failedValidateOperations(String)
+    case unknownError
+  }
+
+  public enum ICloudStatus: String {
+    case available
+    case notSignedIn
+    case noAccess
+  }
+
+  /// The Portal API instance to retrieve the client's and custodian's IDs.
+  public var api: PortalApi?
+  /// The key used to store the private key in iCloud.
+  public var key: String = ""
+
+  public let encryption: PortalEncryption
+
+  private let isSimulator = TARGET_OS_SIMULATOR != 0
+  private let storage: PortalKeyValueStore
+
+  /// Initializes a new ICloudStorage instance.
+  public init(
+    encryption: PortalEncryption? = nil,
+    keyValueStore: PortalKeyValueStore? = nil
+  ) {
+    self.encryption = encryption ?? PortalEncryption()
+    self.storage = keyValueStore ?? PortalKeyValueStore()
+  }
+
+  /*******************************************
+   * Public functions
+   *******************************************/
+
+  public func delete() async throws -> Bool {
+    let key = try await getKey()
+    return self.storage.delete(key)
+  }
+
+  public func read() async throws -> String {
+    let key = try await getKey()
+    return self.storage.read(key)
+  }
+
+  public func write(_ value: String) async throws -> Bool {
+    let key = try await getKey()
+    return self.storage.write(key, value: value)
+  }
+
+  public func validateOperations() async throws -> Bool {
+    let testKey = "portal_test"
+    let testValue = "test_value"
+
+    _ = self.storage.write(testKey, value: testValue)
+
+    let readValue = self.storage.read(testKey)
+    if readValue == testValue {
+      _ = self.storage.delete(testKey)
+      return true
+    } else {
+      throw ICloudStorageError.failedValidateOperations("Failed to read/write test data")
+    }
+  }
+
+  /*******************************************
+   * Private functions
+   *******************************************/
+
+  private func createKey(_ client: ClientResponse) -> String {
+    ICloudStorage.hash("\(client.custodian.id)\(client.id)")
```

Comment:
Why do we call `ICloudStorage` here instead of just `self`

Reply:
Because the `hash` function is static in ICloudStorage. Not sure why, but it works either way. If we want to use `self.` we'd need to make it an instance function, though.